epoxy-adapter/src/main/java/com/airbnb/epoxy/EpoxyVisibilityTracker.kt

```diff
@@ -0,0 +1,506 @@
+                }
+            }
+        }
+    }
+
+    /**
+     * Call this methods every time something related to ui (scroll, layout, ...) or something related
+     * to data changed.
+     *
+     * @param recyclerView        the recycler view
+     * @param epoxyHolder         the [RecyclerView]
+     * @param detachEvent         true if the event originated from a view detached from the
+     * recycler view
+     * @param eventOriginForDebug a debug strings used for logs
+     * @return true if changed
+     */
+    private fun processVisibilityEvents(
+        recyclerView: RecyclerView,
+        epoxyHolder: EpoxyViewHolder,
+        detachEvent: Boolean,
+        eventOriginForDebug: String
+    ): Boolean {
+        if (DEBUG_LOG) {
+            Log.d(
+                TAG,
+                String.format(
+                    "%s.processVisibilityEvents %s, %s, %s",
+                    eventOriginForDebug,
+                    System.identityHashCode(epoxyHolder),
+                    detachEvent,
+                    epoxyHolder.adapterPosition
+                )
+            )
+        }
+        val itemView = epoxyHolder.itemView
+        val id = System.identityHashCode(itemView)
+        var vi = visibilityIdToItemMap[id]
+        if (vi == null) {
+            // New view discovered, assign an EpoxyVisibilityItem
+            vi = EpoxyVisibilityItem(epoxyHolder.adapterPosition)
+            visibilityIdToItemMap.put(id, vi)
+            visibilityIdToItems.add(vi)
+        } else if (epoxyHolder.adapterPosition != RecyclerView.NO_POSITION &&
+            vi.adapterPosition != epoxyHolder.adapterPosition
+        ) {
+            // EpoxyVisibilityItem being re-used for a different adapter position
+            vi.reset(epoxyHolder.adapterPosition)
+        }
+        var changed = false
+        if (vi.update(itemView, recyclerView, detachEvent)) {
+            // View is measured, process events
+            vi.handleVisible(epoxyHolder, detachEvent)
+            partialImpressionThresholdPercentage?.let { percentage ->
+                vi.handlePartialImpressionVisible(
+                    epoxyHolder, detachEvent,
+                    percentage
+                )
+            }
+            vi.handleFocus(epoxyHolder, detachEvent)
+            vi.handleFullImpressionVisible(epoxyHolder, detachEvent)
+            changed = vi.handleChanged(epoxyHolder, onChangedEnabled)
+        }
+        return changed
+    }
+
+    private fun processChildRecyclerViewAttached(childRecyclerView: RecyclerView) {
+        // Register itself in the EpoxyVisibilityTracker. This will take care of nested list
+        // tracking (ex: carousel)
+        var tracker = getTracker(childRecyclerView)
+        if (tracker == null) {
+            tracker = EpoxyVisibilityTracker()
+            tracker.setPartialImpressionThresholdPercentage(partialImpressionThresholdPercentage)
+            tracker.attach(childRecyclerView)
+        }
```

Comment:
We could shorten this a bit by doing:
```suggestion
        val tracker = getTracker(childRecyclerView) ?: EpoxyVisibilityTracker().apply {
            setPartialImpressionThresholdPercentage(partialImpressionThresholdPercentage)
            attach(childRecyclerView)
        }
```